### DISCUSSION TITLE
How to display related fields in a table format?

### DISCUSSION THREAD

**Discussion by Pachat** [directus/directus#10990] [Archive]
At present, the "Display template" of a relational field allows to set the list of the fields
 `{{ name }} {{ quantity_available }} {{ price }}` and put some text in between 
`Name: {{ name }} Available: {{ available_quantity }} Price: {{ price }}`. Nice!

But the different related rows are rendered as
```
<ul>
    <li><span>Name: name1</span><span> Available: available_quantity1</span><span> Price: price1</span></li>
    ...
    <li><span>Name: name3</span><span> Available: available_quantity3</span><span> Price: price3</span></li>
</ul>

```
which is not vertically aligned.

Is there a way to custom the  "Display template" in order to have instead?
```
<table>
    <tr>
        <td>Name: name1</td>
        <td> Available: available_quantity1</td>
        <td> Price: price1</td>
    </tr>
    ...
    <tr>
        <td>Name: name3</td>
        <td> Available: available_quantity3</td>
        <td> Price: price3</td>
    </tr>
</table>

```

**Comment by oreilles** [comment]
It is not possible at the moment. If that is something you really need, consider having a look at the [`related-values`](https://github.com/directus/directus/tree/main/app/src/displays/related-values) display, and writing an [extension](https://docs.directus.io/extensions/displays/) that matches your requirements 👍

**Reply by Pachat** [reply]
As this seems the right way to go, i tried to create the custom display `related-values-table` from `related-values`
First, I installed the doc examples on displays custom components and they work correctly.
Then, I create 
`extensions`
∟  `displays`
&nbsp; &nbsp;     ∟  `related-values-table`
where I put the `package.json`
```
{
  "name": "related-values-table",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@directus/extension-sdk": "^9.0.0-rc.91"
  },
  "directus:extension": {
	  "type": "display",
	  "path": "index.js",
	  "source": "src/index.ts",
	  "host": "^9.0.0-rc.91",
	  "hidden": false
  }
}
```
in `related-values-table`
&nbsp; &nbsp;     ∟  `src`, I put the two files you mentioned [related-values](https://github.com/directus/directus/tree/main/app/src/displays/related-values) 
Now, running `npx directus-extension build` in `related-values-table`, I get this error
```
- Building Directus extension...(node:20612) UnhandledPromiseRejectionWarning: Error: Unexpected token (Note that you need plugins to import files that are not JavaScript)
    at error (C:\xampp\htdocs\node_modules\rollup\dist\shared\rollup.js:158:30)
    at Module.error (C:\xampp\htdocs\node_modules\rollup\dist\shared\rollup.js:12423:16)
    at Module.tryParse (C:\xampp\htdocs\node_modules\rollup\dist\shared\rollup.js:12826:25)
    at Module.setSource (C:\xampp\htdocs\node_modules\rollup\dist\shared\rollup.js:12729:24)
    at ModuleLoader.addModuleSource (C:\xampp\htdocs\node_modules\rollup\dist\shared\rollup.js:22202:20)
```
How can I get rid of this error?

**Reply by joselcvarela** [reply]
As the error says you must need plugins to import files that are not Javascript.
If you create the extension using Javascript but you are using Typescript you need to require the plugins to compile Typescript.

**Reply by Pachat** [reply]
Thank you for pointing this.
So I installed `npm install -g ts-node`
`tsc -v `shows `Version 4.5.5`
But I still get the same error.

**Reply by Pachat** [reply]
Sorry. I tried the whole day and, despite I could make a default display ts component from the installation and could compile it;
it bugs when I replace the index.ts and display.vue with the files [related-values](https://github.com/directus/directus/tree/main/app/src/displays/related-values) 
I hope the table presentation of related values will be added to Directus core.

**Reply by joselcvarela** [reply]
I am not sure if this is related, but `htdocs` is the project name you created via Directus?

Also, do you have any clue @nickrum ?

**Reply by rijkvanzanten** [reply]
@joselcvarela xampp is one of those dev apps that runs Apache and PHP etc (like MAMP on mac), `htdocs` is Apache's root folder 👍🏻

**Reply by nickrum** [reply]
@Pachat You are using a very old version of the Extensions SDK which had no Typescript support. Make sure you scaffold your extension as described [here](https://docs.directus.io/extensions/creating-extensions/).

Also, just copying over the existing `related-values` display will very likely not work because it is not possible to relatively import code from the App. A few dedicated utils and composables are exposed through the Extensions SDK. Anything else has to be copied over.

**Reply by Pachat** [reply]
Understood. So I made a clean scaffold install of the sdk. (displays / typescript)
I could generate properly the default display component.

Now, I inserted the two files [related-values](https://github.com/directus/directus/tree/main/app/src/displays/related-values)
and get these three errors
- '@/utils/adjust-fields-for-displays' is imported by src\index.ts, but could not be resolved – treating it as an external dependency
- '@/utils/get-related-collection' is imported by src\index.ts, but could not be resolved – treating it as an external dependency
- Cannot find module 'sass'
```

npm run build

> directus-extension-related-values-table@1.0.0 build
> directus-extension build

| Building Directus extension...'@/utils/adjust-fields-for-displays' is imported by src\index.ts, but could not be resolved – treating it as an external dependency
'@/utils/get-related-collection' is imported by src\index.ts, but could not be resolved – treating it as an external dependency
'@/stores' is imported by src\index.ts, but could not be resolved – treating it as an external dependency
× Failed

(plugin vue) Error: Cannot find module 'sass'
```

**Reply by nickrum** [reply]
As briefly mentioned above, the process of converting an internal display to a custom one isn't as straightforward. You have to install any missing package, replace imports with those exported from the Extensions SDK and copy over any remaining files.

**Comment by benhaynes** [comment]
All "repeater" style interfaces currently use the display template, but we could add an additional "table" style set of interfaces too (might be too different to make an option of the existing interfaces). We already have a native `v-table` component, so most of the work is already done.

I see the benefit of this and would support the work if someone wanted to take a stab at a PR!

**Reply by Pachat** [reply]
With an option to show or not the columns' header, it would be great!

**Reply by benhaynes** [reply]
Interesting, wouldn't have thought of that option. Could you give a quick use case? Removing the header may also remove certain table functionality... sorting, changing columns, resizing columns, etc.

**Reply by Pachat** [reply]
In practice, I did no find a use case without columns' header, _which we already have with the related field with just one field in the Display template (moreover, if we don't want a header there, we can just set a space as the Field Name Translation)_.

My suggestion was more in the direction to show the columns' header for understanding, and, indeed, very useful for table functionality... sorting, changing columns, resizing columns, etc. as you pointed it out.

**Reply by jamescammarano** [reply]
I'm having trouble imagining how this is different from the current table interface. Would it be a nested table in the table interface?

**Reply by Pachat** [reply]
The current table interface is fine for the data.

But when it comes to **related data** ( M2M for instance ) **within a Display** with a Display Template of **multiple fields**, 
![DisplayTemplateOfRelatedValues](https://user-images.githubusercontent.com/13746133/150646477-64974a65-2ef6-4f34-9f5e-bf1932f4faab.png)
the rendered presentation is done now with comma separated fields. Which is not nice when we have fields of different length.

The goal is to have instead a table presentation of the related fields with the fields aligned vertically.
This is much more readable. _The table presentation of related fields is the only way I use so far in such cases._
I'd like to find this feature with Directus.

@jaycammarano So to answer your question, the nesting wanted is not inside the table interface, but inside the display of an item.

What we have
![Item in Actions](https://user-images.githubusercontent.com/13746133/150651244-99f1ff36-7209-491e-84d2-1a5c9a8d7260.png)

and what we would like to have
![Item in Actions with table](https://user-images.githubusercontent.com/13746133/150651524-24aaf315-c769-4ea9-84ff-c64b2b386fc1.png)

**Reply by jamescammarano** [reply]
Thanks I understand!

**Reply by benhaynes** [reply]
I'm open to looking into adding a "table style" view to the existing relational interfaces, or adding a new "Table" interface to support this. Either way, the issue is just that our core team doesn't have the time for this lower priority update. If someone in the community wants to coordinate a PR, we'd be happy to provide some light guidance along the way! ❤️
